test/unit/pipeline/FetchUnitTest.cc

```diff
@@ -442,5 +458,260 @@ TEST_F(PipelineFetchUnitTest, idleLoopBufferDueToNotTakenBoundary) {
+
+  // Tick
+  fetchUnit.tick();
+
+  // Ensure 2 bytes are consumed
+  EXPECT_EQ(fetchUnit.bufferedBytes_, 0);
+  EXPECT_EQ(fetchUnit.pc_, 16);
+  EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], mOp2);
+}
+
+// Test that invalid half byte held at the end of the buffer is not successfully
+// predecoded and should be retried when reads don't complete
+TEST_F(PipelineFetchUnitTest, invalidHalfWord_readsDontComplete) {
+  ON_CALL(isa, getMaxInstructionSize()).WillByDefault(Return(insnMaxSizeBytes));
+  ON_CALL(isa, getMinInstructionSize()).WillByDefault(Return(insnMinSizeBytes));
+  ON_CALL(memory, getCompletedReads()).WillByDefault(Return(completedReads));
+
+  // Buffer will contain invalid 2 bytes so predecode returns 0 bytes read
+  // TODO is this correct way to not change output
+  ON_CALL(isa, predecode(_, 2, 0xE, _)).WillByDefault(Return(0));
+
+  // getMaxInstructionSize called for second time in assertion
+  if (strcmp(SIMENG_BUILD_TYPE, "Release") == 0) {
+    EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  } else {
+    EXPECT_CALL(isa, getMaxInstructionSize()).Times(2);
+  }
+  EXPECT_CALL(memory, getCompletedReads()).Times(1);
+  EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+  EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+  // Set PC to 14, buffered bytes = 0
+  fetchUnit.updatePC(14);
+  // Tick
+  fetchUnit.tick();
+
+  // No data consumed
+  EXPECT_EQ(fetchUnit.bufferedBytes_, 2);
+  EXPECT_EQ(fetchUnit.pc_, 14);
+  // TODO ensure this is correct
+  EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], MacroOp());
+
+  // Expect that memory is requested even though there is data in the buffer as
+  // bufferedBytes < maxInstructionSize
+  EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  EXPECT_CALL(memory,
+              requestRead(Field(&memory::MemoryAccessTarget::address, 16), _))
+      .Times(1);
+  fetchUnit.requestFromPC();
+
+  EXPECT_EQ(fetchUnit.bufferedBytes_, 2);
+  EXPECT_EQ(fetchUnit.pc_, 14);
+
+  // Memory doesn't complete reads in next cycle but buffered bytes should
+  // attempt to be predecoded
+  ON_CALL(memory, getCompletedReads())
+      .WillByDefault(Return(span<memory::MemoryReadResult>{nullptr, 0}));
+  // Predecode still returns no bytes read
+  ON_CALL(isa, predecode(_, 2, 0xE, _)).WillByDefault(Return(0));
+
+  // getMaxInsnSize called again in assertion
+  if (strcmp(SIMENG_BUILD_TYPE, "Release") == 0) {
+    EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+  } else {
+    EXPECT_CALL(isa, getMaxInstructionSize()).Times(2);
+  }
+  EXPECT_CALL(memory, getCompletedReads()).Times(1);
+  EXPECT_CALL(isa, getMinInstructionSize()).Times(2);
+  EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+  // Tick
+  fetchUnit.tick();
+
+  // Ensure 2 bytes are not consumed
+  EXPECT_EQ(fetchUnit.bufferedBytes_, 2);
+  EXPECT_EQ(fetchUnit.pc_, 14);
+  // TODO ensure this is correct
+  EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], MacroOp());
```

Comment:
I think this is correct yes - as `MacroOp` is an alias for `std::vector<>` thus this will create an empty vector which for this test is what we expect in the output slot

Reply:
TODO removed